### PR TITLE
Allow tapestry import/export/duplicate

### LINKED ISSUE #506: Improve POST Tapestry endpoint to accept JSON bodies
This will allow users to send a JSON representation of a Tapestry and work from there.

### PR DESCRIPTION
**Pull request by narendrasss**
Closes #506 

* Updates the `POST /tapestries` endpoint to allow sending a JSON representation of a Tapestry as the request body
  * This endpoint will create a new Tapestry and copy over the nodes and links to the new Tapestry, assigning them new IDs in the process
  * Returns the newly created Tapestry
* Add an export button to the settings modal that converts the current tapestry to a json file and automatically downloads it

Most of these changes were made so that the Cypress tests may be refactored to be more modular and run more independently. As a result, things like styling and adding an import button is delegated to a future PR.

**Testing**

* Open a Tapestry
* Open the settings modal, then go to the "Advanced" tab
* Click "Export Tapestry"
* A download prompt should open with the filename being `title-of-tapestry.json`
* Open the resulting JSON file and verify the contents look correct

### CODE REVIEW

**Comment by liam-armstrong** [comment]
Export works well, very fast. Interestingly, the JSON file contains my author name, not sure if this is intended based on the description here.
Also, this may be in a different task, but a button to send the JSON file and replace the current Tapestry (a user import function) might be useful.

Otherwise working as expected and ready for merge

**Comment by narendrasss** [comment]
Found some issues, leaving this as a note for myself.

- [ ] Quiz progress does not get reset on export

File: endpoints.php [hunk 1]
```diff
@@ -373,24 +373,51 @@ function getTapestry($request)
 function addTapestry($request)
 {
     $tapestryData = json_decode($request->get_body());
-    if (!get_page_by_title($tapestryData->title, 'OBJECT', 'tapestry')) {
+    try {
+        if (get_page_by_title($tapestryData->title, 'OBJECT', 'tapestry')) {
+            throw new TapestryError('POST_EXISTS');
+        }
```
Comment (wynnset) [inline endpoints.php:379]: I don't think this is required. Users should be allowed to add Tapestries that have the same title.

**Comment by wynnset** [comment]
Oh I just saw that you have fixed the duplicate title thing in your other PR, so you can ignore that part of my review.

**Pull request merged**